internal/memory/pool.go

```diff
@@ -0,0 +1,324 @@
+	}
+
+	// not enough memory in the next class, try to recursively expand
+	if len(nextPool.free) == 0 {
+		err := pa.Expand(nextClsType)
+		if err != nil {
+			return err
+		}
+	}
+
+	offset, err := pa.FindFreeOffset(nextClsType)
+	if err != nil {
+		return err
+	}
+
+	nextReg, err := pa.swap(nextClsType, offset)
+	if err != nil {
+		return err
+	}
+
+	// memCls validity has been checked already, we can ignore the error
+	clsSize, _ := GetMemoryClassSize(clsType)
+
+	// expand memCls
+	cp := pa.pools[clsType]
+	if cp == nil {
+		cp = newEmptyMemoryPool()
+		pa.pools[clsType] = cp
+	}
+	// create 4 smaller regions
+	for i := uint64(0); i < 4; i++ {
+		offset := nextReg.offset + i*clsSize
+		reg := &region{
+			parent: nextReg,
+			class:  clsType,
+			offset: offset,
+		}
+		cp.free[offset] = reg
+	}
+	return nil
+}
+
+// Release marks a memory region of class `memCls` and offset `offset` as free and tries to merge smaller regions into
+// a bigger one
+func (pa *PoolAllocator) Release(memCls classType, offset uint64) error {
+	mp := pa.pools[memCls]
+	if mp == nil {
+		return ErrEmptyPoolOperation
+	}
+
+	n, err := pa.swap(memCls, offset)
+	if err != nil {
+		return err
+	}
+
+	if err := pa.merge(n.parent); err != nil {
+		if err != ErrEarlyMerge {
+			return err
+		}
+	}
+	return nil
+}
+
+func (pa *PoolAllocator) merge(parent *region) error {
+	// nothing to merge
+	if parent == nil {
+		return nil
+	}
+
+	parentCls := parent.class
+	// nothing to do, cannot merge the smallest class
+	if parentCls == 0 {
+		return nil
+	}
+	parentPool := pa.pools[parentCls]
+	// cannot merge an empty pool
+	if parentPool == nil {
+		return ErrEmptyPoolOperation
+	}
+
+	childCls := parentCls - 1
+	childPool := pa.pools[childCls]
+	// no child nodes to merge, try merge parent
+	if childPool == nil {
+		return pa.merge(parent.parent)
+	}
+
+	childSize, err := GetMemoryClassSize(childCls)
+	if err != nil {
+		return err
+	}
+
+	// check if all of the child nodes are free
+	var children []*region
+	for i := uint64(0); i < 4; i++ {
+		child, free := childPool.free[parent.offset+i*childSize]
+		if !free {
+			return ErrEarlyMerge
+		}
+		children = append(children, child)
+	}
+
+	// at this point all of the child nodes will be free and we can merge
+	for _, child := range children {
+		delete(childPool.free, child.offset)
+	}
+
+	if _, err := pa.swap(parent.class, parent.offset); err != nil {
+		return err
+	}
+
+	return pa.merge(parent.parent)
+}
+
+func (pa *PoolAllocator) swap(memCls classType, offset uint64) (*region, error) {
+	clsPool := pa.pools[memCls]
+	if clsPool == nil {
+		return nil, ErrEmptyPoolOperation
+	}
+
+	// check if in busy pool, then mark free
+	if n, ok := clsPool.busy[offset]; ok {
+		clsPool.free[offset] = n
+		delete(clsPool.busy, offset)
+		return n, nil
+	}
+
+	// check if in free pool, then mark busy
+	if n, ok := clsPool.free[offset]; ok {
+		clsPool.busy[offset] = n
+		delete(clsPool.free, offset)
+		return n, nil
+	}
```

Comment:
Should we enforce that only one of these scenarios is valid at a time?

Reply:
split this into 2 different functions.